Add ability to open multiple devices from different processes (libtm and t265)

orshefi opened the pull request:
Sorry for the previous pull request I've done, newbie to working on open source

A direct result of the discussion in #4261 
Here are my changes -passed sanity checks, further tests will be conducted throughout the following week

dorodnic commented:
Hi @orshefi 
Thank you for contributing. We will review and test during upcoming week and provide more feedback

orshefi commented:
Hi @dorodnic,
I developed the solution on ubuntu 18.04 and it works fine there.
Today I've tested it on ubuntu 16.04 and it raises SIGSEGV signal from timer complete function in dispatcher, so far I have no clue why.

Any hints on what to look for? 

Anyhow I think that you should include in your unit tests a test that raises this issue since I passed the tests in TravisCI.

I'm working on a solution meanwhile..

orshefi commented:
@dorodnic - I can see this line everywhere in the Dispatcher:
    ASSERT(mThreadId != std::this_thread::get_id());

orshefi commented:
Found the issue and fixed it in the latest commit:
The constructor of Device registers a timer for the Listener, I am using the constructor several times in order to query all sort of data on the device as part of the DeviceHolder concept (Profile, Intrinsics, Extrinsics, Device Info).

In order to fix it I moved the scheduleTimer function to the start() function in Device class, this function is only called on Device stream start.

Tested to work on both ubuntu 16.04 and 18.04.

orshefi commented:
@ev-mp

radfordi commented:
Hi @orshefi,

Thanks for this PR.   The `libtm` code base can be a challenge to get your head around.   I'm still not there myself.   Can you explain what is the impediment to opening multiple devices from different processes?  It looks to me like it's just the `libusb_claim`.  My understanding is that it's ok to `libusb_open` a device multiple times (from separate libusb contexts either in the same or different processes), but not to `libusb_claim` it.   Have you considered separating these concepts inside `Device` or by using your `DeviceHolder`.  It seems that the `Manager` should never `libusb_claim` except during firmware upload.

radfordi commented:
In particular, it seems to me that open firmware load and `sensor_base::open` should call `libusb_claim`.

dorodnic commented:
@orshefi - sorry for the delay. Both myself and @ev-mp had to be away from GitHub due to unforeseen circumstances in last couple of weeks, but getting this PR merged is high priority for me.

orshefi commented:
Hi guys,

@dorodnic - glad to hear that you see the importance of this feature.

@radfordi - regarding your question:
At first glance, it seemed to us that libusb_claim is the one that locks the devices but after practical testing, we had come to the conclusion that libusb_open is the locking function of the device.

The main purpose of our patch is to split the discovery of the device and profile acquisition and the device opening, we did it by briefly opening the device (hence locking it in DeviceHolder constructor) and then destructed it until the next time that the sensor uses the open function (when pipeline is initiated with a give serial number).

doronhi commented:
@orshefi ,
Thanks for that PR, I would really like to see it merged.
I encountered a problem while testing with realsense-viewer:
In the original version. disconnecting the T265 device while streaming yields the following error message:
```
15:13:18.630 [6432] [E] Device-2830: ERROR: Bulk transfer message 0x2 (DEV_GET_TIME) request to device got LIBUSB_ERROR_IO. Bytes requested 6, bytes transferred 0
15:13:18.630 [6432] [E] Device-2830: Error syncing TM2 and Host clocks
15:13:18.635 [6461] [E] Device-2830: FW crashed - got error in stream endpoint thread function: status = -4 (LIBUSB_ERROR_NO_DEVICE), actual = 0
15:13:18.635 [6460] [E] Device-2830: FW crashed - got error in interrupt endpoint thread function: status = -4 (LIBUSB_ERROR_NO_DEVICE), actual = 0
15:13:18.836 [6432] [E] Device-2830: State [ACTIVE_STATE] got event [ON_ERROR] ==> [ERROR_STATE]
15:13:18.836 [6432] [E] Device-2830: Entered state [ERROR_STATE]
```
With your PR merged, I get the following error message:
```
15:23:39.086 [7843] [E] Device-7AE0: FW crashed - got error in stream endpoint thread function: status = -4 (LIBUSB_ERROR_NO_DEVICE), actual = 0
15:23:39.086 [7842] [E] Device-7AE0: FW crashed - got error in interrupt endpoint thread function: status = -4 (LIBUSB_ERROR_NO_DEVICE), actual = 0
15:23:39.287 [7784] [E] Device-7AE0: State [ACTIVE_STATE] got event [ON_ERROR] ==> [ERROR_STATE]
15:23:39.287 [7784] [E] Device-7AE0: Entered state [ERROR_STATE]
Segmentation fault (core dumped)
```
and the viewer crashes.
Checked on ubuntu 16.04

Also, I ran 2 instances of realsense-viewer. I then disconnected 1 T265 device and It's instance of realsense-viewer crash. Running it again, while the other viewer is still running, yields the following error message in the viewer's screen:

```
Io in rs2_get_stream_profiles(sensor:0x7f983c002420):
Failed to get supported raw streams
```
(hand copied)

It seems related to another test I made: running an instance of the ROS wrapper, realsense2_camera. Disconnecting T265 and then reconnecting it everything is ok but then disconnecting it and reconnecting again, the application crash.

radfordi commented:
@orshefi, does #5213 fix this for you?  If not would you mind thinking about your requirements in the context of #5213, as it is our short-term goal to remove `libtm` completely.

orshefi commented:
That (#5213) seems like a great pull request, it handles all issues I have encountered. 
When does it planned to be released? I won't be able to finish my work on this pull request until February at least.

radfordi commented:
@orshefi, this PR is obsolete now that `libtm` has been removed with merge of #5213.   We maybe still needs something similar, but it would need to be applied directly to `librealsense` and not `libtm` now.  Thanks for all your work on this.
